File: py/leetcode/dynamic_programming.py
# =============================================================================
# Python LeetCode — Dynamic Programming  (Blind 75 / NeetCode 150)
# =============================================================================
from typing import List

# LC #70  Climbing Stairs — Fibonacci O(n) / O(1)
# def climb_stairs(n: int) -> int: pass

# LC #198 House Robber — dp[i] = max(dp[i-1], dp[i-2]+nums[i])
# def rob(nums: List[int]) -> int: pass

# LC #213 House Robber II — run rob() on nums[:-1] and nums[1:]
# def rob_ii(nums: List[int]) -> int: pass

# LC #322 Coin Change — bottom-up O(amount * coins)
# def coin_change(coins: List[int], amount: int) -> int: pass

# LC #518 Coin Change II — count combinations
# def change(amount: int, coins: List[int]) -> int: pass

# LC #300 Longest Increasing Subsequence — O(n²) dp / O(n log n) patience sort
# def length_of_lis(nums: List[int]) -> int: pass

# LC #1143 Longest Common Subsequence — 2D dp O(m*n)
# def longest_common_subsequence(text1: str, text2: str) -> int: pass

# LC #139 Word Break — dp[i] = any(dp[j] and s[j:i] in wordSet)
# def word_break(s: str, word_dict: List[str]) -> bool: pass

# LC #62  Unique Paths — dp[i][j] = dp[i-1][j] + dp[i][j-1]
# def unique_paths(m: int, n: int) -> int: pass

# LC #72  Edit Distance — 2D dp insert/delete/replace
# def min_distance(word1: str, word2: str) -> int: pass

# LC #312 Burst Balloons — interval dp O(n³)
# def max_coins(nums: List[int]) -> int: pass

File: flutter-dart/leetcode/dynamic_programming.dart
// TOPIC: Dynamic Programming LeetCode Problems | dart dynamic_programming.dart

// TODO 1: Climbing Stairs — LC #70
//   int climbStairs(int n)
//   - dp[i] = dp[i-1] + dp[i-2]; optimize to two int variables

// TODO 2: Coin Change — LC #322
//   int coinChange(List<int> coins, int amount)
//   - dp = List.filled(amount+1, amount+1); dp[0]=0

// TODO 3: Longest Increasing Subsequence — LC #300
//   int lengthOfLIS(List<int> nums)
//   - O(n²) dp[i] = max(dp[j]+1) for j<i where nums[j]<nums[i]
//   - O(n log n) patience sorting with binary search

// TODO 4: Longest Common Subsequence — LC #1143
//   int longestCommonSubsequence(String text1, String text2)
//   - 2D dp table; dp[i][j] = if chars match: dp[i-1][j-1]+1 else max(dp[i-1][j], dp[i][j-1])

// TODO 5: 0/1 Knapsack (classic)
//   int knapsack(List<int> weights, List<int> values, int capacity)
//   - dp[w] = max value with capacity w; iterate backwards to avoid reuse

// TODO 6: Decode Ways — LC #91
//   int numDecodings(String s)
//   - dp[i] = ways to decode s[0..i-1]; check 1-digit and 2-digit

// TODO 7: Word Break — LC #139
//   bool wordBreak(String s, List<String> wordDict)
//   - dp[i] = true if s.substring(0,i) can be segmented using wordDict

void main() {}
